### PR TITLE
[Dotenv] Add Dotenv::bootEnv() to check for .env.local.php before calling Dotenv::loadEnv()

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | yes
| Tickets       | -
| License       | MIT
| Doc PR        | -

The goal of this PR is to eventually get rid of the `config/bootstrap.php` file in Symfony 5.1 apps.
I think we've done enough iterations on that piece of bootstrapping logic to put it inside the `Dotenv` component.
This fully replaces https://github.com/symfony/recipes/blob/master/symfony/framework-bundle/4.2/config/bootstrap.php

It doesn't conflict with current apps so they'll be fine keeping the `config/bootstrap.php` file until they're upgraded.

The new bootstrapping logic will require adding this line in `bin/console` and `public/index.php`:
```php
(new Dotenv())->bootEnv(dirname(__DIR__).'/.env');
```

Recipes updated at https://github.com/symfony/recipes/pull/724

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
Recipes updated ready at https://github.com/symfony/recipes/pull/724

File: src/Symfony/Component/Dotenv/Dotenv.php [hunk 1]
```diff
@@ -35,15 +35,47 @@ final class Dotenv
     private $data;
     private $end;
     private $values;
-    private $usePutenv;
+    private $envKey;
+    private $debugKey;
+    private $prodEnvs = ['prod'];
```
Comment (ro0NL) [inline src/Symfony/Component/Dotenv/Dotenv.php]: perhaps for consistency, define `$testEnvs = ['test'] + $defaultEnv = 'dev';` as well? Then cleanup the arg list in `loadEnv()`

Comment (nicolas-grekas) [reply src/Symfony/Component/Dotenv/Dotenv.php]: I considered that, but I think it's not worth the upgrade trouble.

Comment (ro0NL) [reply src/Symfony/Component/Dotenv/Dotenv.php]: it would avoid a half stateful / stateless API, and perhaps is worth it from the component's design POV.

Comment (ro0NL) [reply src/Symfony/Component/Dotenv/Dotenv.php]: do we prefer mutability? or what about `new Dotenv('APP_ENV', 'APP_DEBUG', 'dev', ['prod'], ['test'])` :/

Comment (nicolas-grekas) [reply src/Symfony/Component/Dotenv/Dotenv.php]: We have no issues with mutability when it's purpose isto configure the instance.
Using many constructor args is a pain when you only want to change the last one.
Setters are better here IMHO.

Comment (ro0NL) [reply src/Symfony/Component/Dotenv/Dotenv.php]: i undertand the technical aspect, im curious if a user should change the context between loading envs (vs using a dedicated instance per context). In practice the context is a fixed concern.

Comment (nicolas-grekas) [reply src/Symfony/Component/Dotenv/Dotenv.php]: In practice, people will figure out what they need :)
I considered alternatives before deciding for this design: it looks the best compromise to me.

Comment (ro0NL) [reply src/Symfony/Component/Dotenv/Dotenv.php]: IMHO it's component design first :) i strongly feel the _right_ design is

```php
$dotenv = new Dotenv('APP_ENV', 'APP_DEBUG', 'dev', ['prod'], ['test']);
$dotenv->load('.env', $putenv = true);
$dotenv->loadAll(['.env'], $putenv = true);
```

but im fine either way, my only use case is a generated one :)

Comment (nicolas-grekas) [reply src/Symfony/Component/Dotenv/Dotenv.php]: I understand, that's what I meant by "compromise". We're not designing from the ground up, and in practice this is the most seamless.

**Review by lyrixx — APPROVED** [review]

**Comment by nicolas-grekas** [comment]
ping @symfony/mergers

File: src/Symfony/Component/Dotenv/Dotenv.php [hunk 2]
```diff
@@ -35,15 +35,47 @@ final class Dotenv
...
 
     /**
-     * @var bool If `putenv()` should be used to define environment variables or not.
-     *           Beware that `putenv()` is not thread safe, that's why this setting defaults to false
+     * @param string $envKey
      */
-    public function __construct(bool $usePutenv = false)
+    public function __construct($envKey = 'APP_ENV', string $debugKey = 'APP_DEBUG')
+    {
+        if (\in_array($envKey = (string) $envKey, ['1', ''], true)) {
```
Comment (yceruto) [inline src/Symfony/Component/Dotenv/Dotenv.php]: `is_bool($envKey)` ?

**Review by yceruto — APPROVED** [review]

File: src/Symfony/Component/Dotenv/Dotenv.php (hunk 2 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/Dotenv/Dotenv.php]: wouldn't make a diff, and this style makes the block "removable" as is in 6.

**Review by chalasr — APPROVED** [review]

**Comment by chalasr** [comment]
Rebase needed

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**